kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt

```diff
@@ -210,25 +210,53 @@ class RequestInterpreter(private val schemaModel: SchemaModel) {
+            copy(selections = selections + node)
+        } ?: this
+
+        // https://spec.graphql.org/October2021/#sec-Unions
+        //  "With interfaces and objects, only those fields defined on the type can be queried directly; to query
+        //  other fields on an interface, typed fragments must be used. This is the same as for unions, but unions
+        //  do not define any fields, so *no* fields may be queried on this type without the use of type refining
+        //  fragments or inline fragments (with the exception of the meta-field `__typename`)."
         val unionMembersChildren: Map<Type, List<Execution>> =
             field.returnType.possibleTypes.associateWith { possibleType ->
                 val selections = selectionNode.selectionSet?.selections
+                // Add __typename to all fragment selections if specified
+                val typenameSelection = selections?.firstOrNull { it is FieldNode && it.name.value == "__typename" }
 
-                val a = selections?.filterIsInstance<FragmentSpreadNode>()?.firstOrNull {
+                val namedFragment = selections?.filterIsInstance<FragmentSpreadNode>()?.firstOrNull {
                     ctx.fragments[it.name.value]?.first?.name == possibleType.name
                 }
 
-                if (a != null) return@associateWith handleReturnType(
-                    ctx,
-                    possibleType,
-                    ctx.fragments.getValue(a.name.value).second
-                )
+                if (namedFragment != null) {
+                    return@associateWith handleReturnType(
+                        ctx,
+                        possibleType,
+                        ctx.fragments.getValue(namedFragment.name.value).second.addSelectionNode(typenameSelection)
+                    )
+                }
 
-                val b = selections?.filterIsInstance<InlineFragmentNode>()?.find {
+                val inlineFragment = selections?.filterIsInstance<InlineFragmentNode>()?.find {
                     possibleType.name == it.typeCondition?.name?.value
                 }
 
-                if (b != null) return@associateWith handleReturnType(ctx, possibleType, b.selectionSet)
+                if (inlineFragment != null) {
+                    return@associateWith handleReturnType(
+                        ctx,
+                        possibleType,
+                        inlineFragment.selectionSet.addSelectionNode(typenameSelection)
+                    )
+                }
+
+                // At this point only __typename was requested
+                if (typenameSelection != null) {
+                    return@associateWith handleReturnType(
+                        ctx,
+                        possibleType,
+                        selectionNode.selectionSet!!.addSelectionNode(typenameSelection)
+                    )
+                }
 
                 throw ValidationException(
```

Comment:
Also not sure if clients actually _have to_ provide a selection for _all_ types.

Reply:
https://graphql.com/learn/interfaces-and-unions/#querying-with-unions does allow it but from my understanding that exception is correct. Types without fragment result in an empty selection set, and according to the spec, empty selection sets are invalid. Will keep as-is.